src/Workspaces/Core/Portable/Storage/SQLite/v2/SQLitePersistentStorageConstants.cs

```diff
@@ -61,37 +66,51 @@ internal static class SQLitePersistentStorageConstants
         ///
-        /// This gives a very efficient integral key, and means that the we only have to store a
-        /// single mapping from stream name to ID in the string table.
-        ///
         /// The format of the table is:
         ///
+        ///  <code>
         ///  ProjectData
-        ///  -------------------------------------------------------------------
-        ///  | DataId (primary key, integer) | | Checksum (blob) | Data (blob) |
-        ///  -------------------------------------------------------------------
+        ///  ------------------------------------------------------------------------------------------------
+        ///  | ProjectPathId (int) | ProjectNameId (int) | DataNameId (int) | Checksum (blob) | Data (blob) |
+        ///  ------------------------------------------------------------------------------------------------
+        ///  | Primary Key                                                  |
+        ///  ----------------------------------------------------------------
+        ///  </code>
         /// </summary>
         public const string ProjectDataTableName = "ProjectData" + Version;
 
         /// <summary>
-        /// Inside the DB we have a table for data that we want associated with a <see cref="Document"/>.
-        /// The data is keyed off of an integral value produced by combining the ID of the Document and
-        /// the ID of the name of the data (see <see cref="SQLitePersistentStorage.ReadStreamAsync(DocumentKey, Document?, string, Checksum?, CancellationToken)"/>.
-        ///
-        /// This gives a very efficient integral key, and means that the we only have to store a
-        /// single mapping from stream name to ID in the string table.
+        /// Inside the DB we have a table for data that we want associated with a <see cref="Document"/>. The data is
+        /// keyed off the project information, and the folder and name of the document itself.  This allows the majority
+        /// of the key to be shared (project path/name, and folder name) with other documents, and only having the doc
+        /// name portion be distinct.  Different TFM flavors will also share everything but the project name.
         ///
         /// The format of the table is:
         ///
+        ///  <code>
         ///  DocumentData
-        ///  -------------------------------------------------------------------
-        ///  | DataId (primary key, integer) | | Checksum (blob) | Data (blob) |
-        ///  -------------------------------------------------------------------
+        ///  ------------------------------------------------------------------------------------------------------------------------------------------------
+        ///  | ProjectPathId (int) | ProjectNameId (int) | DocumentFolderId (int) | DocumentFolderId (int) | DataNameId (int) | Checksum (blob) | Data (blob) |
```

Comment:
DocumentFolderId is in there twice
Also - why the document folder?  If the project info is all the same, what difference does having a diff document folder make?

Reply:
> DocumentFolderId is in there twice

oops, second should be DocumentNameId.

> If the project info is all the same, what difference does having a diff document folder make?

A major difference.  Imagine one single project that contains `Foo\Bar.cs` and `Baz\Bar.cs`.  Those are two files with the same name (but differnet folders).

Reply:
Could that be accomplished by using just a single document file path instead of name + folder?

Reply:
> Could that be accomplished by using just a single document file path instead of name + folder?

Yes.  but at the cost of more space.  Imagine you have the following (not hypothetical):

```
C:\github\roslyn\src\Workspaces\Core\Portable\Workspace\Workspace.cs
C:\github\roslyn\src\Workspaces\Core\Portable\Workspace\Workspace_Editor.cs
``

Then using filepath means both of those strings are fully stored in the DB.  with folder+name, we store the following 3 strings instead:

```
C:\github\roslyn\src\Workspaces\Core\Portable\Workspace
Workspace.cs
Workspace_Editor.cs
```

Something we can/should look into is if we can instead compute the relative filepath from the project path, and the relative projectpath from the solutionpath.  Which would save a lot more.

Reply:
i'm going to make that change in this PR.